src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -0,0 +1,1054 @@
+#include "expression/list_expression.hpp"
+#include "expression/logical_expression.hpp"
+#include "expression/pqp_column_expression.hpp"
+#include "expression/pqp_select_expression.hpp"
+#include "expression/value_expression.hpp"
+#include "expression_functors.hpp"
+#include "like_matcher.hpp"
+#include "operators/abstract_operator.hpp"
+#include "resolve_type.hpp"
+#include "scheduler/current_scheduler.hpp"
+#include "sql/sql_query_plan.hpp"
+#include "storage/materialize.hpp"
+#include "storage/value_column.hpp"
+#include "utils/assert.hpp"
+#include "utils/lambda_visitor.hpp"
+
+using namespace std::string_literals;
+using namespace opossum::expression_factory;
+
+namespace opossum {
+
+ExpressionEvaluator::ExpressionEvaluator(const std::shared_ptr<const Table>& table, const ChunkID chunk_id)
+    : _table(table), _chunk(_table->get_chunk(chunk_id)) {
+  _output_row_count = _chunk->size();
+  _column_materializations.resize(_chunk->column_count());
+}
+
+template <typename R>
+std::shared_ptr<ExpressionResult<R>> ExpressionEvaluator::evaluate_expression_to_result(
+    const AbstractExpression& expression) {
+  switch (expression.type) {
+    case ExpressionType::Arithmetic:
+      return _evaluate_arithmetic_expression<R>(static_cast<const ArithmeticExpression&>(expression));
+
+    case ExpressionType::Logical:
+      return _evaluate_logical_expression<R>(static_cast<const LogicalExpression&>(expression));
+
+    case ExpressionType::Predicate:
+      return _evaluate_predicate_expression<R>(static_cast<const AbstractPredicateExpression&>(expression));
+
+    case ExpressionType::Select: {
+      const auto* pqp_select_expression = dynamic_cast<const PQPSelectExpression*>(&expression);
+      Assert(pqp_select_expression, "Can only evaluate PQPSelectExpression");
+      return _evaluate_select_expression<R>(*pqp_select_expression);
+    }
+
+    case ExpressionType::Column: {
+      const auto* pqp_column_expression = dynamic_cast<const PQPColumnExpression*>(&expression);
+      Assert(pqp_column_expression, "Can only evaluate PQPColumnExpressions");
+      return _evaluate_column_expression<R>(*pqp_column_expression);
+    }
+
+    // ValueExpression and ParameterExpression both need to unpack an AllTypeVariant, so one functions handles both
+    case ExpressionType::Parameter:
+    case ExpressionType::Value:
+      return _evaluate_value_or_parameter_expression<R>(expression);
+
+    case ExpressionType::Function:
+      return _evaluate_function_expression<R>(static_cast<const FunctionExpression&>(expression));
+
+    case ExpressionType::Case:
+      return _evaluate_case_expression<R>(static_cast<const CaseExpression&>(expression));
+
+    case ExpressionType::Exists:
+      return _evaluate_exists_expression<R>(static_cast<const ExistsExpression&>(expression));
+
+    case ExpressionType::Extract:
+      return _evaluate_extract_expression<R>(static_cast<const ExtractExpression&>(expression));
```

Comment:
Is `extract` really a separate expression or is it only a function with a weird syntax?

Reply:
Would have loved for it to be a normal function, but didn't know how to get the `MONTH, DAY, YEAR, HOUR` part into an Expression that could be passed in. If more functions using enum parameters pop up, we should merge them. Until then Extract is its own Expression

Reply:
Maybe copy this into a comment